### PR TITLE
Move associated items to their own page.

### PR DESCRIPTION
**Pull request by Havvy**
Because they are a cutting concern between impls and traits, they don't really fit in either.

If anybody has concrete suggestions for how to make this better, I'm all ears. I'm not a fan of a lot of my writing in this PR. Especially the definitions.

### CODE REVIEW

File: src/items/associated-items.md [hunk 1]
```diff
@@ -0,0 +1,260 @@
+# Associated Items
+
+*Associated Items* are the items declared in [traits] or defined in
+[implementations]. They are called this because they are defined on an associate
+type &mdash; the type in the implementation. They are a subset of the kinds of
+items you can declare in a module. Specifically, there are [associated
```
Comment (alercah) [inline src/items/associated-items.md]: Perhaps "associated functions (including methods)"?

File: src/items/associated-items.md [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+associating item. For example, the `is_some` method on `Option` is intrinsically
+related to Options, so should be associated.
+
+Associated items are also the contract that traits have. 
+
+## Associated functions and methods
+
+*Associated functions* are [functions] associated with a type.
+
+An *associated function declaration* is written as `fn`, then an [identifier]
```
Comment (alercah) [inline src/items/associated-items.md]: This applies throughout the doc: spelling grammar out explicitly in words seems verbose and hard to read. Moreover, this seems redundant with regular function grammar. Do we have a BNF standard we can use? Or perhaps update the grammar page and link to that?

File: src/items/associated-items.md [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+    fn new() -> Struct {
+        Struct {
+            field: 0i32
+        }
+    }
+}
+```
+
+When the associated function is declared on a trait, the function can be called
+on the trait. When this happens, it is substituted for
```
Comment (alercah) [inline src/items/associated-items.md]: "on both the trait and on any type implementing the trait"?

File: src/items/associated-items.md [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+    fn from_i32(n: i32) -> Self;
+}
+impl Num for f64 {
+    fn from_i32(n: i32) -> f64 { n as f64 }
+}
+let x: f64 = Num::from_i32(42);
+let x: f64 = f64::from_i32(42);
+```
+
+Associated functions whose first parameter is named `self` are called *methods*
```
Comment (alercah) [inline src/items/associated-items.md]: I feel like methods should be their own shorthand, perhaps?

File: src/items/associated-items.md [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+and may be invoked using the [method call operator], for example, `x.foo()`, as
+well as the usual function call notation.
+
+When the first parameter is named `self`, the following shorthands may be used:
```
Comment (alercah) [inline src/items/associated-items.md]: Are lifetimes permitted in shorthands? e.g. `&'a self`

File: src/items/associated-items.md [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+An *associated type declaration* is written as `type`, then an [identifier], and
+finally an optional trait bounds.
+
+The identifier is the name of the declared type alias. The optional trait bounds
+must be fulfilled by the implementations of the type alias.
+
+An *associated type definition* is written as `type`, then an [identifier], then
+an `=`, and finally a [type].
+
+If an item `Item` has an associated type `Assoc`, then `Item::Assoc` is a type
```
Comment (alercah) [inline src/items/associated-items.md]: Does `Self` count as a built-in associated type? I feel like it should be mentioned here.

File: src/items/associated-items.md [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+
+An *associated constant declaration* is written as `const`, then an identifier,
+then `:`, then a type, finished by a `;`.
+
+The identifier is the name of the constant used in the path. The type is the
+type that the definition has to implement.
+
+An *associated constant definition* is written as a declaraction, but between
+the type and the `;`, there is an `=` followed by a [constant expression].
+
```
Comment (alercah) [inline src/items/associated-items.md]: Perhaps add more discussion of how to refer to associated constants, as you did with types? e.g. `Item::Const`

File: src/items/functions.md [hunk 8]
```diff
@@ -12,10 +12,10 @@ on completion.
 [type]: types.html
 
 When referred to, a _function_ yields a first-class *value* of the
-corresponding zero-sized [*function item type*][function item type], which
+corresponding zero-sized [*function item type*], which
 when called evaluates to a direct call to the function.
 
-[function item type]: types.html#function-item-types
+[*function item type*]: types.html#function-item-types
```
Comment (alercah) [inline src/items/functions.md]: Move the formatting outside the link makes more sense to me, perhaps. I wasn't even aware this link syntax worked with formatting.

File: src/items/associated-items.md [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+# type Surface = i32;
+# type BoundingBox = i32;
+trait Shape {
+    fn draw(&self, Surface);
+    fn bounding_box(&self) -> BoundingBox;
+}
+```
+
+This defines a trait with two methods. All values that have [implementations]
+of this trait in scope can have their `draw` and `bounding_box` methods called.
```
[the remark this answers is not in the document]
Comment (Havvy) [reply src/items/associated-items.md]: You misread; it's the traits that have scope (visibility?), not the implementations. Either way, gonna have to reword this to avoid that confusion.

**Comment by Havvy** [comment]
Well, took me long enough, but I've addressed feedback where I can.

**Comment by Havvy** [comment]
The TravisCI failure is not against the latest change. Seems I might have confused TravisCI with multiple pushes to my branch. >_>

File: src/items/associated-items.md [hunk 10]
```diff
@@ -0,0 +1,272 @@
...
+    fn empty() -> Vec<T> { Vec::new() }
+    fn insert(&mut self, x: T) { self.push(x); }
+}
+```
+
+## Associated Constants
+
+*Associated constants* are [constants] associated with a type.
+
+An *associated constant declaration* declares a specificatoin for associated
```
Comment (matthewjasper) [inline src/items/associated-items.md]: specificatoin -> specification

File: src/items/associated-items.md [hunk 11]
```diff
@@ -0,0 +1,272 @@
...
+constant definitions. It is written as `const`, then an identifier,
+then `:`, then a type, finished by a `;`.
+
+The identifier is the name of the constant used in the path. The type is the
+type that the definition has to implement.
+
+An *associated constant definition* defines a constant associated with another
```
Comment (matthewjasper) [inline src/items/associated-items.md]: a type

**Comment by Havvy** [comment]
Updated.

**Comment by matthewjasper** [comment]
Thanks!
